process_report/institute_list_models.py

```diff
@@ -0,0 +1,61 @@
+from typing import Annotated
+import datetime
+
+import pydantic
+import validators
+
+
+def parse_date(v: str) -> str:
+    try:
+        datetime.datetime.strptime(v, "%Y-%m")
+        return v
+    except ValueError:
+        raise ValueError(f"Invalid date string {v}. Must be in format YYYY-MM")
+
+
+def validate_domain(v: str) -> str:
+    if not validators.domain(
+        v, consider_tld=True
+    ):  # Ensures only TLDs allowed by IANA are valid
+        raise ValueError(f"Invalid domain {v}")
+    return v
+
+
+DateField = Annotated[str, pydantic.BeforeValidator(parse_date)]
+DomainField = Annotated[str, pydantic.AfterValidator(validate_domain)]
+
+
+class InstituteInfo(pydantic.BaseModel):
+    display_name: str
+    domains: list[DomainField]
+    mghpcc_partnership_start_date: DateField = None
+    include_in_nerc_total_invoice: bool = None
```

Comment:
Similar to the above, although in this case you could instead write:

```suggestion
    include_in_nerc_total_invoice: bool = False
```

Reply:
Pydantic should have yelled at you if you attempted to instantiate these objects using default values. If it didn't, maybe we need to add an additional test.

Reply:
I just tested `Pydantic` too, and it seems if the field is not present in the YAML file, `Pydantic` is fine with the field defaulting to `None`. I'm fine writing a unit test for the validator, or a validator for the validator.

Reply:
I guess it's just type checkers (like mypy or basedpyright) that complain:

```
  /home/lars/redhat/moc/invoicing/process_report/institute_list_models.py:31:48 - error: Type "None" is not assignable to declared type "DateField"
    "None" is not assignable to "str" (reportAssignmentType)
```